packages/language-server/test/snapshots/xml-view-diagnostics/utils.ts

```diff
@@ -0,0 +1,106 @@
+import { resolve, sep } from "path";
+import { Diagnostic } from "vscode-languageserver-types";
+import { TextDocument } from "vscode-languageserver";
+import { readJsonSync, readFileSync } from "fs-extra";
+import { expect } from "chai";
+import { sortBy, forEachRight } from "lodash";
+import { generateModel } from "@ui5-language-assistant/test-utils";
+import { UI5SemanticModel } from "@ui5-language-assistant/semantic-model-types";
+import { getXMLViewDiagnostics } from "../../../src/xml-view-diagnostics";
+
+export const INPUT_FILE_NAME = "input.xml";
+export const OUTPUT_LSP_RESPONSE_FILE_NAME = "output-lsp-response.json";
+export const OUTPUT_RANGES_FILE_NAME = "output-ranges.xml";
+
+export function getExpectedDiagnosticsLSPResponse(
+  testDir: string
+): Diagnostic[] {
+  const sourcesTestDir = toSourcesTestDir(testDir);
+  const lspResponsePath = resolve(
+    sourcesTestDir,
+    OUTPUT_LSP_RESPONSE_FILE_NAME
+  );
+  const expectedDiagnostics = readJsonSync(lspResponsePath);
+  return expectedDiagnostics;
+}
+
+const ui5ModelPromise = generateModel({ version: "1.71.14" });
+let ui5Model: UI5SemanticModel | undefined = undefined;
+
+export async function getActualDiagnosticLSPResponse(
+  testDir: string
+): Promise<Diagnostic[]> {
+  // No top level await
+  ui5Model = await ui5ModelPromise;
+  const sourcesTestDir = toSourcesTestDir(testDir);
+  const inputPath = resolve(sourcesTestDir, INPUT_FILE_NAME);
+  const xmlSnippetContents = readFileSync(inputPath).toString("utf-8");
+  const xmlTextDoc = TextDocument.create(
+    `file://${inputPath}`,
+    "xml",
+    0,
+    xmlSnippetContents
+  );
+  const actualDiagnostics = getXMLViewDiagnostics({
+    document: xmlTextDoc,
+    ui5Model,
+  });
+  return actualDiagnostics;
+}
+
+export async function getActualXMLWithRIssueRangesMarked(
+  testDir: string
+): Promise<string> {
+  const sourcesTestDir = toSourcesTestDir(testDir);
+  const inputPath = resolve(sourcesTestDir, INPUT_FILE_NAME);
+  const xmlSnippetContents = readFileSync(inputPath).toString("utf-8");
+
+  const xmlTextDoc = TextDocument.create(
+    `file://${inputPath}`,
+    "xml",
+    0,
+    xmlSnippetContents
+  );
+
+  const actualDiagnostics = await getActualDiagnosticLSPResponse(testDir);
+  // this assumes no overlapping ranges
```

Comment:
This is fine for the current tests/implemented diagnostics but I'm not sure we can keep this assumption in the future.
It can be improved in another PR, but I thought we could keep a list of position and insert character (which would be one of the arrows) and sort that, and then do the same logic you did here (go over the list backwards) but add the insert character instead of both arrows.

Reply:
Yeah, we can upgrade it **if** needed. I'm not yet sure if we need to, lets see after we implement more validations